File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-3.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*, Node* > map; 
        Node* q = head; 
        Node dummy(0);
        Node* p = &dummy; 
        while(q){
            if(map.find(q) == map.end()){
                Node* newNode = new Node(q->val);
                p->next = newNode;
                map[q] = newNode;
            }
            else{
                p->next = map[q];
            }
            p = p->next;
            if(q->random == nullptr){
                p->random =nullptr;
            }
            else if(map.find(q->random) == map.end()){
                Node* newNode =new Node(q->random->val);
                p->random = newNode;
                map[q->random] = newNode;
            }
            else{
                p->random = map[q->random];
            }
            q = q->next;
        }
        return dummy.next; 
    }
};

File: Data Structures & Algorithms/add-two-numbers/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        head = dummy
        rem = 0
        while l1 or l2:
            num = 0
            if l1 and l2: 
                num = l1.val + l2.val + rem
                l1 = l1.next
                l2 = l2.next
            elif l1: 
                num = l1.val + rem
                l1 = l1.next
            else: 
                num = l2.val + rem
                l2 = l2.next
            tmp = ListNode(num%10)
            if num >= 10:
                rem = num//10
            else: 
                rem = 0
            dummy.next = tmp
            dummy = dummy.next
        if rem > 0: 
            tmp2 = ListNode(rem)
            dummy.next = tmp2
        return head.next